runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -356,9 +522,103 @@ impl module::MigrationHandler for Module {
+        // fees will be returned in do_evm above, once the execution is
+        // finished and we have calculated the actual fees.
+        // In addition, we will also withdraw Oasis fees for this tx from
+        // the same EVM account.
+
+        let oasis_fees = tx.auth_info.fee.amount.amount();
+        let evm_fees: u128 = match tx.call.method.as_str() {
+            "evm.Withdraw" => 0,
+            "evm.Call" => {
+                let call: CallTx = cbor::from_value(tx.call.body.clone()).map_err(|_| {
+                    CoreError::MalformedTransaction(format_err!(
+                        "evm.Call call body doesn't decode into CallTx"
+                    ))
+                })?;
+                // Make sure that gas_price >= min_gas_price.
+                if call.gas_price < min_gas_price {
+                    return Err(CoreError::MalformedTransaction(anyhow::Error::new(
+                        Error::GasPriceTooLow,
+                    )));
+                }
+                let evm_fee = call
+                    .gas_price
+                    .checked_mul(U256::from(call.gas_limit))
+                    .ok_or(CoreError::GasOverflow)?
+                    .checked_add(call.value)
+                    .ok_or(CoreError::GasOverflow)?;
+                evm_fee.as_u128()
+            }
+            "evm.Create" => {
+                let create: CreateTx = cbor::from_value(tx.call.body.clone()).map_err(|_| {
+                    CoreError::MalformedTransaction(format_err!(
+                        "evm.Create call body doesn't decode into CreateTx"
+                    ))
+                })?;
+                // Make sure that gas_price >= min_gas_price.
+                if create.gas_price < min_gas_price {
+                    return Err(CoreError::MalformedTransaction(anyhow::Error::new(
+                        Error::GasPriceTooLow,
+                    )));
+                }
+                let evm_fee = create
+                    .gas_price
+                    .checked_mul(U256::from(create.gas_limit))
+                    .ok_or(CoreError::GasOverflow)?
+                    .checked_add(create.value)
+                    .ok_or(CoreError::GasOverflow)?;
+                evm_fee.as_u128()
+            }
+            _ => return Err(CoreError::AuthenticateTxNotHandled),
+        };
 
-impl module::BlockHandler for Module {
+        let total_fees: u128 = oasis_fees + evm_fees;
+
+        // Take the tokens from the EVM account and move them into the
+        // fee accumulator.
+        let state = ctx.runtime_state();
+        let store = storage::PrefixStore::new(state, &MODULE_NAME);
+        let mut accounts =
+            storage::TypedStore::new(storage::PrefixStore::new(store, &state::ACCOUNTS));
+        let mut account: evm_backend::Account = accounts.get(&evm_acct_addr).unwrap_or_default();
+
+        if account.balance < total_fees.into() {
+            return Err(CoreError::InsufficientFeeBalance);
```

Comment:
So currently, you must always first deposit enough into the EVM account to cover any gas fees for the EVM invocations, right?

Reply:
With the recent change, should this return not authenticated as maybe the core account has enough funds?